### PR TITLE
Tpetra: Fix BlockCrs unit test on CUDA.

### PR DESCRIPTION
**Pull request by kyungjoo-kim**
## Description

The new comm interface in Tpetra::BlockCrs fails unit tests on CUDA architectures. This PR address the problem. Using DualView, it is very very very tricky to track where it is modified or synced. 

Especially, when we pass dual views through virtual interface, one can expect polymophic behaviors based on the derived classes. In such a case, input dual view should be synced and both device and host buffers are available for the function.
```
/// callee function can safely assume that inputs are available for both host and device
/// callee function make sure output view modify flags
///   - when this function is virtual, some function may modify host side and some may modify device 
void CalleeFunction(const DualView<value_type*>& in_a, 
                                   DualView<value_type*> &out_b,
                                   ArrayView<value_type> &out_c) {
  // callee function does not sync inputs but check if the data is available or not 
  assert(!in_a.need_sync_device());
  // this function can choose whether it uses host algorithm or device algorithm. e.g.,
  out_b.modify_device();
  parallel_for(device_range_policy, []() { doSomethingDevice(in_a.view_device(), out_b.view_device(); );  
  // input can be also used for host algorithm. 
   assert(!in_a.need_sync_host()); 
   parallel_for(host_range_policy, []() { doSomethingHost(in_a.view_host(), out_c); });
}
void CallerFunction() {
  DualView<value_type*> in_a;
  DualView<value_type*> out_b;
  DualView<value_type*> out_c;
  // make sure in_a is synced for both
  in_a.sync_host(); in_a.sync_device();
  // out_b is output and its modification flag will be adjusted in the callee function
  // out_c host view is extracted from the dual view; its modification flag should be adjusted in the caller
  auto out_c_av = getArrayViewFromHost(out_c);
  out_c.modify_host();
  CalleeFunction(in_a, out_b, out_c_av);
}
```
Maybe all people who use the dualviews already follow the above practice. It is something new to me. 

## Related Issues

#4257  #4162 

## How Has This Been Tested?

Following the instruction described in #4257, I reproduced the error and confirm that all unit tests are passed with this PR.

### CODE REVIEW

**Comment by bartlettroscoe** [comment]
@kyungjoo-kim, given the files that got modified in this PR, this turned out to be a defect in the library code and not the test code?

**Comment by mhoemmen** [comment]
@kyungjoo-kim Would you like a merge right away to fix this?

**Comment by mhoemmen** [comment]
@kddevin wrote:
> They sound as if there is a deeper issue in the packAndPrepareNew / doTransfer design; is that correct?

There are _several_ issues in the design ;-)  @crtrott and I did some rough benchmarking in the Tpetra CG solve with 2 GPUs on waterman, and found a few:

  1. Export and Import need to store those four LID arrays (permute to, permute from, remotes, and exports) as DualView.
  2. `MultiVector::unpackAndCombine` creates DualViews on each call.  Get rid of `castAwayConstDualView` and syncs there.

That doesn't even touch on Distributor's "slow path."

**Comment by kyungjoo-kim** [comment]
@kddevin I am not sure if it is a design problem or not. My point was that when we pass a dual view as an input argument, we have a certain assumption about which buffer is going to be used. The function needs to be understood on the context where it is used to avoid unnecessary dual view sync. In addition, reading the code is quite tricky to track modification of a dual view if its usage is different for different developers. So, I just thought it might be better if we have a common assumption when dual views are used as function arguments. 

@mhoemmen I first merged this PR first. I will respond to your comment later and submit additional PR if it is necessary to account for your comments. Thanks.

File: packages/tpetra/core/src/Tpetra_DistObject_decl.hpp [hunk 1]
```diff
@@ -824,6 +824,9 @@ namespace Tpetra {
                        const Kokkos::DualView<const local_ordinal_type*, device_type>& /* exportLIDs */,
                        Kokkos::DualView<packet_type*, buffer_device_type>& /* exports */,
                        const Kokkos::DualView<size_t*, buffer_device_type>& /* numPacketsPerLID */,
+                       /// Kyungjoo: numPacketsPerLID is better remove const. its content is modified and
```
[the remark this answers is not in the document]
Comment (kyungjoo-kim) [reply packages/tpetra/core/src/Tpetra_DistObject_decl.hpp]: Yes. As it is const, if I modify the buffer, I need to use ``const_cast`` to raise a modify flag.

File: packages/tpetra/core/src/Tpetra_DistObject_def.hpp [hunk 2]
```diff
@@ -1861,9 +1889,40 @@ namespace Tpetra {
...
-                                     this->numImportPacketsPerLID_,
-                                     constantNumPackets, distor, CM);
+          
+          // Kyungjoo: remoteLIDs is read only and it should be synced 
+          //   before doTransfer is called (see line 581).
+          //   We still need to sync inputs.
+          //   The commOnHost is a very tricky thing. This variable 
+          //   can control where communication happens and where data
+          //   is updated. However, when we pass dual views to a function,
+          //   we do not have a control where data is available.
```
[the remark this answers is not in the document]
Comment (kyungjoo-kim) [reply packages/tpetra/core/src/Tpetra_DistObject_def.hpp]: I agree.

**Comment by mhoemmen** [comment]
@kyungjoo-kim I made progress on the fixes for extra DualView initializations this weekend.  Hoping to be done in a day or two.

**Pull request merged**